io/src/main/scala/fs2/io/file/Watcher.scala

```diff
@@ -288,45 +302,50 @@ object Watcher {
       }
 
     override def events(pollTimeout: FiniteDuration): Stream[F, Event] =
-      unfilteredEvents(pollTimeout).zip(registrations.continuous).flatMap {
+      unfilteredEvents(pollTimeout).evalMap(e => registrations.get.map((e, _))).flatMap {
         case ((key, events), registrations) =>
-          val reg = registrations.get(key)
-          val filteredEvents = reg
-            .map(reg =>
-              events.filter(e =>
-                reg.eventPredicate(e) && !(e
-                  .isInstanceOf[Event.Created] && reg.suppressCreated)
-              )
-            )
-            .getOrElse(Nil)
+          val regs = registrations.filter(_.key == key)
+          val filteredEvents = events.filter { e =>
+            regs.foldLeft(false) { (acc, reg) =>
+              acc || {
+                val singleFileMatch = reg.singleFile && Event.pathOf(e) == Some(reg.path)
+                val dirMatch =
+                  !reg.singleFile && !(e.isInstanceOf[Event.Created] && reg.suppressCreated)
+                singleFileMatch || dirMatch
+              }
+            }
```

Comment:
This looks like `exists` to me. Would that be clearer?

Reply:
It also saves some allocations.